plugins/promela/hu.bme.mit.gamma.promela.verification/src/hu/bme/mit/gamma/promela/verification/TraceBackAnnotator.xtend

```diff
@@ -0,0 +1,277 @@
+	protected final Logger logger = Logger.getLogger("GammaLogger")
+	
+	new(Package gammaPackage, Scanner traceScanner) {
+		this(gammaPackage, traceScanner, true)
+	}
+	
+	new(Package gammaPackage, Scanner traceScanner, boolean sortTrace) {
+		this.gammaPackage = gammaPackage
+		this.traceScanner = traceScanner
+		this.sortTrace = sortTrace
+		this.component = gammaPackage.firstComponent
+		this.promelaQueryGenerator = new PromelaQueryGenerator(component)
+		this.xStsBackAnnotator = new XstsBackAnnotator(promelaQueryGenerator, PromelaArrayParser.INSTANCE)
+		val schedulingConstraintAnnotation = gammaPackage.annotations
+				.filter(SchedulingConstraintAnnotation).head
+		if (schedulingConstraintAnnotation !== null) {
+			this.schedulingConstraint = schedulingConstraintAnnotation.schedulingConstraint
+		}
+		else {
+			this.schedulingConstraint = null
+		}
+	}
+	
+	def ExecutionTrace execute() {
+		// Creating the trace component
+		val trace = createExecutionTrace => [
+			it.component = this.component
+			it.import = this.gammaPackage
+			it.name = this.component.name + "Trace"
+		]
+		val topComponentArguments = gammaPackage.topComponentArguments
+		// Note that the top component does not contain parameter declarations anymore due to the preprocessing
+		checkState(topComponentArguments.size == component.parameterDeclarations.size, 
+			"The number of top component arguments and top component parameters are not equal: " +
+				topComponentArguments.size + " - " + component.parameterDeclarations.size)
+		logger.log(Level.INFO, "The number of top component arguments is " + topComponentArguments.size)
+		trace.arguments += topComponentArguments.map[it.clone]
+		var step = createStep
+		trace.steps += step
+		
+		// Parsing
+		var backAnnotatorState = BackAnnotatorState.INIT
+		var modelState = ModelState.INIT
+		var lastModelState = ModelState.INIT
+		var traceState = TraceState.NOT_REQUIRED
+		var initError = false // error in INIT
+		try {
+			while (traceScanner.hasNext) {
+				var line = traceScanner.nextLine.trim // Trimming leading white spaces
+				switch (modelState) {
+					case INIT: {
+						switch (line) {
+							case line.endsWith(INIT_START): {
+								// Adding reset
+								step.actions += createReset
+								backAnnotatorState = BackAnnotatorState.STATE_CHECK
+								
+								traceState = TraceState.REQUIRED
+								line = traceScanner.nextLine.trim
+								//println("INIT")
+							}
+							case line.equals(INIT_END): {
+								traceState = TraceState.NOT_REQUIRED
+								modelState = ModelState.ENV
+							}
+							// error in INIT state
+							case line.startsWith(TRACE_END): {
+								// Adding reset
+								step.actions += createReset
+								backAnnotatorState = BackAnnotatorState.STATE_CHECK
+								
+								traceState = TraceState.REQUIRED
+								initError = true
+								line = traceScanner.nextLine.trim
+								//println("INIT")
```

Comment:
Delete these commented println-s, or leave them uncommented while using a logger (to help debugging).